### PR TITLE
Add oasis-login authentication through ssh keys

### PR DESCRIPTION
**Pull request by DrDaveD**
See [SOFTWARE-3961](https://opensciencegrid.atlassian.net/browse/SOFTWARE-3961)

Includes changes from #124, merge that first.

### CODE REVIEW

File: goc/share/oasis-batch-worker
Comment (edquist) [inline goc/share/oasis-batch-worker]:
```suggestion
       if [ -n "$CILOGONID" ]; then
```

or equivalently, just

```suggestion
       if [[ $CILOGONID ]]; then
```

File: goc/share/oasis-batch-worker
Comment (edquist) [inline goc/share/oasis-batch-worker]: I know this wasn't your change from this PR, but i'll note it here that all the `[ ... ]` tests with unquoted variables would be safer written with `[[ ... ]]`, which does not do word splitting and glob pathname expansion.

```suggestion
       elif [[ ! -f $cert ]] ; then
```

File: goc/share/generate_sshauthkeys
Comment (edquist) [inline goc/share/generate_sshauthkeys]:
```suggestion
cmd = 'curl -s --cert /etc/grid-security/hostcert.pem --key /etc/grid-security/hostkey.pem "https://topology.opensciencegrid.org/oasis-managers/json?vo=*"'
```

If you are going to use a single command string, and you want to quote the `*` because it's a glob character, well, _technically_ `?` is a glob character too, so you might as well quote the whole url string.

Alternatively you can avoid the shell interpretation question by splitting the cmd args yourself:

```suggestion
cmd = 'curl -s --cert /etc/grid-security/hostcert.pem --key /etc/grid-security/hostkey.pem https://topology.opensciencegrid.org/oasis-managers/json?vo=*'.split()
```

(Unless the whole point is to have bash (with -x) display the curl cmdline)

File: goc/share/generate_sshauthkeys
Comment (edquist) [inline goc/share/generate_sshauthkeys]: Just curious, why the `bufsize=1` ?

File: goc/share/generate_sshauthkeys
Comment (edquist) [reply goc/share/generate_sshauthkeys]: Also, just a note, i think `curl -s` is completely silent, even on errors.  You can use `curl -sS ...` to keep error messages but not display regular informational messages.

File: goc/share/generate_sshauthkeys
Comment (edquist) [inline goc/share/generate_sshauthkeys]: Probably these will work fine if the users' homedirs do not contain whitespace or shell metacharacters, but if you want to be safe about it you can use `subprocess.call`, eg

```suggestion
                subprocess.call(['diff', authkeysfile, authkeysfile + '.new'])
```

File: goc/share/oasis-batch-worker
Comment (DrDaveD) [reply goc/share/oasis-batch-worker]: Oh dear, I guess I didn't test that.  I wrote it awhile ago and forgot I hadn't tested it.  Thanks for catching it.

File: goc/share/generate_sshauthkeys
Comment (DrDaveD) [reply goc/share/generate_sshauthkeys]: Thanks for asking, because it was a copy/paste error from another application.  I will delete that.

File: goc/share/oasis-batch-worker
Comment (DrDaveD) [reply goc/share/oasis-batch-worker]: I'll fix those unquoted variables.

File: goc/share/generate_sshauthkeys
Comment (DrDaveD) [reply goc/share/generate_sshauthkeys]: I'll use -sS

**Comment by DrDaveD** [comment]
Ok I have tested the changes for writing to update.details, found a couple more bugs, and fixed them.  I also applied most of Carl's suggestions.  Please re-review.

**Comment by edquist** [comment]
@DrDaveD, @brianhlin suggested i confirm with you -- the current behavior of the oasis-managers endpoint is only to return oasis manager info for contacts that have a CILogonID (which currently is only you and matyasselmeci).  Is that OK or does your script expect to see oasis manager info for contacts without a CILogonID also?  (If the latter, @brianhlin suggested we may want to hold off on merging until we get everyone registered)

**Comment by edquist** [comment]
@brianhlin has suggested that we can have the endpoint return all oasis managers, and omit the cilogon stuff for the ones without a CILogonID. (We could for instance return CILogonID: null, or we could omit the CILogonID and sshPublicKeys attributes altogether.)

**Comment by DrDaveD** [comment]
The new script only needs those with CILogonID.  The old script is still running using a different topology endpoint in order to make a grid-mapfile for gsissh.  So it's fine for us to put this into production as it is, and we'll eventually turn off the other script.

On the other hand, there is a third script, which creates the accounts.  It is using the same URL as the grid-mapfile script:
```
https://my.opensciencegrid.org/vosummary/xml?summary_attrs_showoasis=on&all_vos=on&active=on&active_value=1&oasis=on&oasis_value=1&sort_key=name
```
That does need all the VOs that have oasis enabled.  If you would rather stop supporting that URL and have the script use the new one, then it would need the list of VOs beyond just those that have CILogonIDs.

**Comment by DrDaveD** [comment]
@edquist Can you please re-review?

File: goc/share/generate_sshauthkeys
Comment (edquist) [inline goc/share/generate_sshauthkeys:13]: Note that the curl command-line will be printed to stderr, presumably that's what you want

File: goc/share/generate_sshauthkeys
Comment (edquist) [inline goc/share/generate_sshauthkeys]: Probably it's fine how you have it, but letting the shell interpret strings that you didn't provide or sanitize always makes me nervous

```suggestion
                subprocess.call(['diff', authkeysfile, authkeysfile + '.new'])
```

File: goc/share/oasis-batch-worker
Comment (edquist) [inline goc/share/oasis-batch-worker:20]: Just a minor note, but for variable assignments like this (`vo=$1`), the quotes are not actually necessary.

File: goc/etc/cron.d/cvmfs
Comment (edquist) [inline goc/etc/cron.d/cvmfs:4]: Presumably the `2>/dev/null` is to discard `sem` errors when as many snapshot jobs as cores are already running.  But (in case it was not your intention), note that you are now also discarding any stderr from the cvmfs server snapshot command also.

**Review by edquist — APPROVED** [review]
A couple notes, which do what you like with, but otherwise thumbs-up

**Comment by edquist** [comment]
PS, @DrDaveD, as it was discussed earlier in this ticket - the topology oasis-managers/json endpoint now returns oasis manager info for all oasis managers, whether or not they have a CILogonID.  (CILogonID and sshPublicKeys will just be omitted if the contact has no CILogonID.)

File: goc/share/generate_sshauthkeys
Comment (DrDaveD) [reply goc/share/generate_sshauthkeys:13]: Yes it is.

File: goc/etc/cron.d/cvmfs
Comment (DrDaveD) [reply goc/etc/cron.d/cvmfs:4]: It's ok, the snapshot -a option sends all the output including errors to log files.

**Comment by DrDaveD** [comment]
It appears edquist must not be on the list of approvers because it's still not letting me merge.  @zvada please also approve.

**Pull request merged**
